ramips: Add support for Beeline SmartBox TURBO+

MaxS0niX opened the pull request:
# [Beeline Smart Box TURBO+](https://openwrt.org/inbox/toh/beeline/smartbox_turbo_plus)
This PR adds support for router Beeline Smart Box TURBO+.
## Specifications
- SoC Type: MediaTek MT7621AT (880 MHz, 2 cores)
- RAM (Nanya NT5CC64M16GP): 128 MiB
- Flash (Macronix MX30LF1G18AC): 128 MiB
- Wireless 2.4 GHz (MT7603EN): b/g/n, 2x2
- Wireless 5 GHz (MT7615N): a/n/ac, 4x4
- Ethernet: 5 ports - 5×GbE (WAN, LAN1-4)
- USB ports: 1xUSB3.0
- Buttons: 2 button (reset, wps)
- LEDs: Red, Green, Blue
- Zigbee (EFR32MG1B232GG): 3.0
- Stock bootloader: U-Boot 1.1.3
- Power: 12 VDC, 1.5 A
### Installation
* Web UI Update
  1. Rename the firmware so that there are no dots in the name, except
  	for the dot in the ".img" file type. Example "openwrt-factory.img".
  2. Upload and update the firmware via the original web interface.

Remark: You might need make the 2rd step twice if your running firmware
is booted from the Slot 1 (Sercomm0 bootflag). The stock firmware
reverses the bootflag (Sercomm0 / Sercomm1) on each firmware update.
#### Discussion forum topic: 
- [Add support for Beeline SmartBox Turbo+](https://forum.openwrt.org/t/add-support-for-beeline-smartbox-turbo/99635?u=maxs0nix).
##### Related devices (SerComm OEM/ODM):
* Sercomm S1500 - #4770 
  * [Beeline SmartBox PRO](https://wikidevi.wi-cat.ru/Beeline_SmartBox_Pro)
  * [WiFire S1500.NBN](https://wikidevi.wi-cat.ru/WiFire_S1500.NBN)
* Sercomm S2
  * [Beeline SmartBox GIGA](https://forum.openwrt.org/t/add-support-for-beeline-smartbox-giga-on-stock-u-boot/99390) - :heavy_check_mark: #4195
* Sercomm S3
  * [Beeline SmartBox TURBO+](https://forum.openwrt.org/t/add-support-for-beeline-smartbox-turbo/99635) - #4108
  * [Beeline SmartBox TURBO](https://forum.openwrt.org/t/add-support-for-beeline-smartbox-turbo/99635/37) - :heavy_check_mark: #10118
  * [Etisalat Sercomm S3](https://forum.openwrt.org/t/add-support-for-sercomm-s3-on-stock-uboot/95970) - #11352
  * Rostelecom RT-SF-1 - :heavy_check_mark: #11137
  * Rostelecom RT-FE-1
###### Previous
PR #3435 (by author Andrew Freeman labz56@gmail.com)

adschm commented on target/linux/ramips/dts/mt7621_beeline_smartbox-turbo-plus.dts:
```diff
@@ -0,0 +1,213 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/dts-v1/;
```
dts-v1 can be dropped here, it's in the dtsi.

adschm commented on target/linux/ramips/dts/mt7621_beeline_smartbox-turbo-plus.dts:
```diff
@@ -0,0 +1,213 @@
+			gpios = <&gpio 13 GPIO_ACTIVE_HIGH>;
+			trigger-sources = <&xhci_ehci_port1>, <&ehci_port2>;
+			linux,default-trigger = "usbport";
+			label = "smartbox-turbo-plus:blue:status";
```
Please use color:function only, i.e. blue:status etc.

adschm commented on target/linux/ramips/dts/mt7621_beeline_smartbox-turbo-plus.dts:
```diff
@@ -0,0 +1,213 @@
+};
+
+&gmac0 {
+	mtd-mac-address = <&factory 0x4>;
```
Nothing in 0x28, 0x2e, 0xe000, 0xe006 ...?

kar200 replied:
Hi @MaxS0niX ,

I have an similar device from my ISP. Thansk for the work you are doing here. 

From what I saw on the 4pda forum the mac address should be in "&factory 0x21000". Do you get the correct IP address?

You should also increment the wirleess by 2 and 3 for example as this is what is done in my device stock firmware.

MaxS0niX replied:
> Hi @MaxS0niX ,
> 
> I have an similar device from my ISP. Thansk for the work you are doing here.
> 
> From what I saw on the 4pda forum the mac address should be in "&factory 0x21000". Do you get the correct IP address?
> 
> You should also increment the wirleess by 2 and 3 for example as this is what is done in my device stock firmware.

What do you have for the device? 
Device get the correct MAC for all interfaces.
MAC LAN = AA:BB:CC:DD:EE:F**2** or AA:BB:CC:DD:EE:F**C**
MAC WAN = MAC LAN+1
MAC Wi-Fi (2.4 GHz) = MAC LAN
MAC Wi-Fi (5 GHz) = MAC LAN+2

adschm replied:
Please don't resolve issues when they are not resolved. The result of this discussion is that we don't know which MAC address assignment is correct.

Closing unresolved issues will look like you try to hide shortcomings of your commit (and is very annoying for, because for people like this I have to reopen anything and check what else should be hidden from my view ...)

MaxS0niX replied:
> Please don't resolve issues when they are not resolved. The result of this discussion is that we don't know which MAC address assignment is correct.
> 
> Closing unresolved issues will look like you try to hide shortcomings of your commit (and is very annoying for me, because for people like this I have to reopen anything and check what else should be hidden from my view ...)

Sorry, I thought this was resolved in the past [PR](https://github.com/openwrt/openwrt/pull/3435#discussion_r491474100).

MaxS0niX replied:
> Nothing in 0x28, 0x2e, 0xe000, 0xe006 ...?

In "Factory" i looked at three devices:
```
0x4 		- LAN MAC
0x8004		- LAN MAC +2 (Wi-Fi 5 GHz)
0x21000		- LAN MAC
0x21010		- Serial number on the backplate stick
0x21050		- ?Version
0x21060		- PCBASN - Serial number on PCB stick
0x21080		- SSID 2G
0x210A0		- PASS 2G

No adresses
0xe000	- FF FF FF FF FF FF
0xe006 	- FF FF FF FF FF FF
0x28 	- FD 9B FF FF FF FF
0x2e 	- FF FF FF FF FF FF
```
If required, I can send an image of mtd2.

MaxS0niX replied:
Command output # sc_fl_map in U-boot:
```
Name                Offset      Length      
calibration1        200000      1000        
calibration2        208000      1000        
mac                 221000      6           
CSN                 221010      c           
language ID         221040      4           
Domain ID           221050      4           
PCBASN              221060      c           
PIN                 221070      8           
SSID                221080      20          
Passphrase          2210a0      40          
```

adschm commented:
Please squash all changes together into a single proper commit (and wrap the long lines in the commit message) ...

MaxS0niX commented:
> Please squash all changes together into a single proper commit (and wrap the long lines in the commit message) ...

Fixed.

kar200 reviewed (approved):
Sorry not sure why this is here. Set as resolved

csharper2005 reviewed (approved):
It works. Tested on my SmartBox Turbo+.

adschm commented:
So, now you have a merge commit again ...

MaxS0niX commented:
> So, now you have a merge commit again ...

Fix. Github offered to resolve the conflict.

csharper2005 commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -187,6 +187,25 @@ define Device/asus_rt-n56u-b1
+  BLOCKSIZE := 128k
+  PAGESIZE := 2048
+  KERNEL_SIZE := 4096k
+  IMAGE_SIZE := 124416k
```
The main purpose of IMAGE_SIZE is for check-size, and here it is used by rootfs0.bin. So it has to be adjusted to 120320k.

MaxS0niX replied:
Fixed.

adschm commented:
Your commit message needs some clean up, e.g.:
* Use your full proper name for Signed-off-by
* Add headlines and spaces between blocks to enhance readability
* Don't just add a big list of links. Reduce it to a few useful ones and tell what they are good for

adschm commented on target/linux/ramips/dts/mt7621_beeline_smartbox-turbo-plus.dts:
```diff
@@ -0,0 +1,210 @@
+		groups = "uart2", "jtag";
+		function = "gpio";
+	};
+};
```
Missing newline at EOF.

MaxS0niX replied:
Fixed.

adschm commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -187,6 +187,25 @@ define Device/asus_rt-n56u-b1
+  DEVICE_VENDOR := Beeline
+  DEVICE_MODEL := SmartBox TURBO+
+  DEVICE_PACKAGES := kmod-mt7603 kmod-mt7615e kmod-mt7615-firmware kmod-usb3 \
+  kmod-usb-ledtrig-usbport uboot-envtools
```
one single tab for hanging indent.

MaxS0niX replied:
Fixed.

MaxS0niX commented:
#### Update: 
- BootLoader: U-boot Sercomm (OEM)
- Script for Header Image (Thanks @kar200 & @csharper2005 )
- Add partion for U-boot Environment
- Reset counter for boot_count1

adschm commented:
> MaxS0niX requested a review from adschm 4 days ago

With the current amount of open PRs, I don't see any reason to review work-in-progress more than once.

hanipouspilot commented on target/linux/ramips/mt7621/base-files/etc/init.d/bootcount:
```diff
@@ -8,6 +8,9 @@ boot() {
 			echo -e "bootcount\nbootchanged\n" | /usr/sbin/fw_setenv -s -
 		;;
+	beeline,smartbox-turbo-plus)
+		echo -ne "$(echo FF | sed 's/../\\x&/g')" | dd of=/dev/mtdblock4 bs=1 seek=$((0x20001)) count=1
```
This should be
```printf '\xff' | dd of=/dev/mtdblock4 bs=1 seek=$((0x20001)) count=1

MaxS0niX replied:
Better to do a check at the beginning.
```[ "$(dd if=/dev/mtd4 skip=$((0x20001)) count=1 bs=1 2>/dev/null)" = "$(printf '\xff')" ] || printf '\xff' | dd of=/dev/mtdblock4 seek=$((0x20001)) bs=1 count=1```

hanipouspilot replied:
What is the point of this check? And I see no pint in using printf to compare.
```
[[ $(dd if=/dev/mtd4 skip=$((0x20001)) count=1 bs=1 2>/dev/null) -eq 0xff ]]
```

MaxS0niX replied:
Check if the counter has not incremented, then there is nothing to do. Useful for those who changed the bootloader to another. 

Have you checked?
```
root@openwrt:~# [[ $(dd if=/dev/mtd4 skip=$((0x20001)) count=1 bs=1 2>/dev/null)
 -eq 0xff ]] && echo 0 || echo 1
ash: �: out of range
1
```

hanipouspilot replied:
This syntax is the last thing we need to worry about now. I'll fix it later.

MaxS0niX replied:
Fixed.

hanipouspilot commented:
There is a way to keep the original flash layout and install Openwrt without soldering UART and bridges. Dual boot feature can be also kept. I've tested it and it seems to work. There is still a problem with validation of the installed firmware by the bootloader. After 1 or 2 reboots it switches back to the stock firmware. That's needs to be resolved.

Generally this seems to be the correct way for an official Openwrt support. This custom layout can be suggested as another option, like it was done with DIR-825b1 on ar71xx. It was called "fat" image.

But 32 MB is enough for most of users. So keeping dual boot is feasible.

csharper2005 commented:
> After 1 or 2 reboots it switches back to the stock firmware. That's needs to be resolved.

How is it possible? Stock rootfs2 is squashed in current layout. So, even the router switches to second kernel for any reason it can't boot because rootfs2 checksum is wrong. Afterwards, it switches back to the first kernel with openwrt.

hanipouspilot commented:
> > After 1 or 2 reboots it switches back to the stock firmware. That's needs to be resolved.
> 
> How is it possible? Stock rootfs2 is squashed in current layout. So, even the router switches to second kernel for any reason it can't boot because rootfs2 checksum is wrong. Afterwards, it switches back to the first kernel with openwrt.

I don't know how it is possible, but this is what is happening.

csharper2005 commented:
> I don't know how it is possible, but this is what is happening.

Ok. Did you compile using exactly the code from this commit and flash using instruction from the commit message?

hanipouspilot commented:
> > I don't know how it is possible, but this is what is happening.
> 
> Ok. Did you compile using exactly the code from this commit and flash using instruction from the commit message?

I used the OEM branch. And flashed it from the stock firmware using mtd_write. I think it is the best way, but we need to solve this problem first. The code in this commit requires soldering, changes the layout and kills dual boot. I don't like it as an official build.

csharper2005 commented:
> I used the OEM branch. And flashed it from the stock firmware using mtd_write. I think it is the best way, but we need to solve this problem first.

Ok, you use different branch, different layout and another flash instruction where the problem exists. But you have no uart to check the boot log. Is it correct? The problem description is not enough in this case I think.

Let me disagree regarding dual-boot. The common best practice is using big ubi instead of tiny. Look at Xiaomi nand ramips device.

MaxS0niX commented:
> I used the OEM branch. And flashed it from the stock firmware using mtd_write. I think it is the best way, but we need to solve this problem first. The code in this commit requires soldering, changes the layout and kills dual boot. I don't like it as an official build.

Please tell us or show how to make sysupgrade from one partition layout to another. It is very interesting.
If you really want an OEM partition, then you need to make the update a file from the web interface. We will be happy to help.
Most users want to use most of the NAND Flash memory. Therefore, I would like to keep both options in this case.

hanipouspilot commented:
>  But you have no uart to check the boot log. Is it correct?

I will solder bridges next week and see what is happening.


>  The common best practice is using big ubi instead of tiny. Look at Xiaomi nand ramips device.

The common best practice is to install Openwrt without soldering if possible. If it is not an option, then there is no other way. Also generally the practice is to keep the OEM layout too. there may be exceptions. But IMHO in this case it makes no sense.

hanipouspilot commented:
> Most users want to use most of the NAND Flash memory.

How do you know what "most users" want? Are you sure they want to open the case and solder bridges and UART?  This is not feasible for a common user. For those who really need a huge rootfs, can have a custom layout. Also why not have both ways?

csharper2005 commented:
> The common best practice is to install Openwrt without soldering if possible. 

You're right. Uart is only an option. But very reliable and universal, e.g. for the devices with bad erase blocks on nand. And it's the best choice for the commit.

Hence, I think that the best place for the full list of the possible flash options is wiki, not a commit message.

hanipouspilot commented:
> Hence, I think that the best place for the full list of the possible flash options is wiki, not a commit message.

This commit can be flashed only using UART. It is a no-go for a common user. So if there is an alternative way, it should be used as the default way. The bad-block problem is another issue.

csharper2005 commented:
> This commit can be flashed only using UART. It is a no-go for a common user. So if there is an alternative way, it should be used as the default way. The bad-block problem is another issue

1. Uart is a normal practice for the commits. I can give you examples if you want.
2. Uart is the only option that works with 100%.

MaxS0niX commented:
> > Most users want to use most of the NAND Flash memory.
> 
> How do you know what "most users" want? Are you sure they want to open the case and solder bridges and UART? This is not feasible for a common user. For those who really need a huge rootfs, can have a custom layout. Also why not have both ways?

Users from [4pda.ru](https://4pda.to/forum/index.php?showtopic=943587). You are having a squabble. I don’t mind the changes, but I will change the main commit to an untested way. I told you above, if you want an OEM partition, let's make an update together from the web interface. Tell us about sysupgrade with other partitions. And both versions of OEM & Big NAND will work, the user does not have to solder, the firmware will be in one action. Stop making discussions for no reason. Let's get to work.

MaxS0niX commented:
0f96aac - Update:
- Added check for the counter.
- Updated sercomm-kernel.sh from #4195

MaxS0niX commented:
@adschm Can you please tell me how you can promote the adoption of the patch[ [RFT] ramips: use partition table on Netgear NAND devices](http://patchwork.ozlabs.org/project/openwrt/patch/20200628232747.1367531-1-jan@3e8.eu/)? This patch is required for this device and GIGA, used by [Sercomm Part Map](https://forum.openwrt.org/t/add-support-for-beeline-smartbox-turbo/99635/28).

hanipouspilot commented:
>  Can you please tell me how you can promote the adoption of the patch

As Adrian said in a comment to the patch, it should be split into two parts.

csharper2005 commented:
> > Can you please tell me how you can promote the adoption of the patch
> 
> As Adrian said in a comment to the patch, it should be split into two parts.

If somebody decides to review it... We are missing a reviewer.

hanipouspilot commented:
> If somebody decides to review it... We are missing a reviewer.

Adrian wrote a useful comment. This patch can be resent as a patch only to the parser.

janh commented:
> Adrian wrote a useful comment. This patch can be resent as a patch only to the parser.

The most recent version of the patch, which splits it up into parser and device changes, is here: https://lists.openwrt.org/pipermail/openwrt-devel/2021-February/034033.html

If you look at the replies, there was the suggestion to upstream the parser. I did not continue to work on that that until now. If anyone wants to do this, please feel free to do so!

(And yes, additional static partition offsets in the partition definition should not be needed, but I didn't test this.)

hanipouspilot commented:
> If you look at the replies, there was the suggestion to upstream the parser. I did not continue to work on that that until now. If anyone wants to do this, please feel free to do so!

Upstreaming may be a good idea, but it will take a very long time to get to Openwrt from there. But we badly need it here, because many devices are unusable with fixed partitions. it is a shame that it seems that the Openwrt maintainers are not interested, or don't have these devices.

MaxS0niX commented:
Ping @adschm

MaxS0niX commented:
Updated the repository, deleted the commit and GitHub closed the PR. 
I tried to restore according to the instructions:
- https://openwrt.org/docs/guide-developer/working-with-github-pr#reopen_closed_pr
It did not work, the "Reopen pull request" button is grayed out and does not allow pressing it.

csharper2005 commented:
@adschm , @mpratt14 hi! 

The instruction https://openwrt.org/docs/guide-developer/working-with-github-pr#reopen_closed_pr didn't help in @MaxS0niX 's case. 

Do you have any additional idea what else they could try to reopen the PR?

MaxS0niX commented:
I wrote a request to GitHub Support, 2 days no response.
Sorry to distract you, @adschm or anyone on the OpenWRT team, I need your help. I need to reopen the PR, I restored the repository to the last commit, but the "Reopen pull request" button is not active.

mpratt14 commented:
unfortunately, this is another rare case where github support is the only decent answer

from what I understand, what needs to happen is a force push to the git object `refs/pull/4108/HEAD`, but it belongs to the openwrt repository and none of us have write permission to it

maybe Adrian can *attempt* what I'm talking about but probably they don't want to experiment...

maybe try another way to contact github like Twitter DM?

mpratt14 commented:
nevermind, I just read that the refs are read-only

https://docs.github.com/en/github/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/checking-out-pull-requests-locally#error-failed-to-push-some-refs

xabolcs commented:
@MaxS0niX 

How about [this gist](https://gist.github.com/robertpainsi/2c42c15f1ce6dab03a0675348edd4e2c) about reopening?
And [this](https://gist.github.com/robertpainsi/2c42c15f1ce6dab03a0675348edd4e2c#gistcomment-3555095) comment?

I think you should reposition your `beeline-smartbox-turbo` branch to commit `cdc646c03282b91d7e9518bb726f29c1de5d24fc` (where this PR - `refs/pull/4108/head` points to) and add a change (e.g. cherry-picking your old commit).

Let's say your branch is on the old commit ... :
```
$ git branch -v
* beeline-smartbox-turbo             4f062d7c0e ramips: Add support for Beeline SmartBox TURBO+
$ git fetch origin refs/pull/4108/head
From https://github.com/openwrt/openwrt
 * branch                  refs/pull/4108/head -> FETCH_HEAD
$ git log -1 --oneline FETCH_HEAD 
cdc646c032 tools/libtool: update to 2.4.2
$ git reset --hard FETCH_HEAD 
HEAD is now at cdc646c032 tools/libtool: update to 2.4.2
$ git cherry-pick 4f062d7c0e
Auto-merging target/linux/ramips/image/mt7621.mk
Auto-merging target/linux/ramips/image/Makefile
Auto-merging package/boot/uboot-envtools/files/ramips
[beeline-smartbox-turbo 57fcb9d770] ramips: Add support for Beeline SmartBox TURBO+
 Author: Maximilian Weinmann <x1@disroot.org>
 Date: Tue Jul 13 15:18:07 2021 +0700
 7 files changed, 296 insertions(+), 3 deletions(-)
 create mode 100755 scripts/sercomm-kernel.sh
 create mode 100644 target/linux/ramips/dts/mt7621_beeline_smartbox-turbo-plus.dts
$ git log -3 --oneline beeline-smartbox-turbo 
57fcb9d770 (HEAD -> beeline-smartbox-turbo) ramips: Add support for Beeline SmartBox TURBO+
cdc646c032 tools/libtool: update to 2.4.2
ae33ce2af2 bcm53xx: specify switch ports for more devices
```

I think pushing that 57fcb9d770211f1d870a0b85899a74076da8cc59 commit to your `beeline-smartbox-turbo` branch would reopen this PR.

MaxS0niX commented:
@mpratt14, @xabolcs thanks for answers)

> maybe try another way to contact github like Twitter DM?

I am not registered on Twitter.

> And [this](https://gist.github.com/robertpainsi/2c42c15f1ce6dab03a0675348edd4e2c#gistcomment-3555095) comment?

I tried to make one more commit from the top, but the button was still inactive, now I will try different variations. 
Thanks for the links.

> $ git fetch origin refs/pull/4108/head
> From https://github.com/openwrt/openwrt
>  * branch                  refs/pull/4108/head -> FETCH_HEAD

```
❯ git fetch origin refs/pull/4108/head
fatal: couldn't find remote ref refs/pull/4108/head
```
Is such a link to FETCH possible?

xabolcs commented:
> Is such a link to FETCH possible?

Looks like your `origin` is not `https://github.com/openwrt/openwrt.git`. You could list your remotes with: `git remote -v show`!

Try direct url instead:
```
$ git fetch https://github.com/openwrt/openwrt.git refs/pull/4108/head
From https://github.com/openwrt/openwrt
 * branch                  refs/pull/4108/head -> FETCH_HEAD
```

MaxS0niX commented:
> You could list your remotes with: `git remote -v show`!

Now I understand.)
```
❯ git remote -v show
origin	git@github.com:MaxS0niX/openwrt.git (fetch)
origin	git@github.com:MaxS0niX/openwrt.git (push)
repa	https://gitlab.com/openwrt-dev-ru-smartbox/openwrt-mt7621-sercomm-smartbox.git (fetch)
repa	https://gitlab.com/openwrt-dev-ru-smartbox/openwrt-mt7621-sercomm-smartbox.git (push)
upstream	https://github.com/openwrt/openwrt.git (fetch)
upstream	https://github.com/openwrt/openwrt.git (push)
```
"Repa" is a repository for tests. And it was necessary to change origin to upstream.

MaxS0niX commented:
> I think pushing that 57fcb9d770211f1d870a0b85899a74076da8cc59 commit to your `beeline-smartbox-turbo` branch would reopen this PR.

I tried your instructions. Did not work out. https://github.com/MaxS0niX/openwrt/commits/beeline-smartbox-turbo
```
❯ git log -3 --oneline beeline-smartbox-turbo
29323edd7b (HEAD -> beeline-smartbox-turbo) ramips: Add support for Beeline SmartBox TURBO+
cdc646c032 (upstream/master, origin/master, origin/HEAD, master) tools/libtool: update to 2.4.2
ae33ce2af2 bcm53xx: specify switch ports for more devices
❯ git push -f origin beeline-smartbox-turbo            
To github.com:MaxS0niX/openwrt.git
 + 4f062d7c0e...29323edd7b beeline-smartbox-turbo -> beeline-smartbox-turbo (forced update)
```
![Снимок экрана_2021-09-10_17-43-22](https://user-images.githubusercontent.com/22564088/132842053-3a68e55a-96a2-4387-8cc6-c269ed8a631f.png)

xabolcs commented:
What happens when you force push commit cdc646c03282b91d7e9518bb726f29c1de5d24fc first and then force push again to 29323edd7b8cd973c78ea1456f8621cb5b8b0bda?

MaxS0niX commented:
> What happens when you force push commit [cdc646c](https://github.com/openwrt/openwrt/commit/cdc646c03282b91d7e9518bb726f29c1de5d24fc) first and then force push again to [29323ed](https://github.com/openwrt/openwrt/commit/29323edd7b8cd973c78ea1456f8621cb5b8b0bda)?

My repository was pointing to 4f062d7 --> 29323ed. [Log](https://privatebin.at/?9fad7b7e00abcd05#Fs8BoadsWiUkqT3CuU7ntU5QfjM3G2b4gaP1JDu7Xee1)
...
Perhaps it took some time or something else to change. But now the button is active! Support has not yet responded to the ticket. Thank you all for your help! : D

MaxS0niX commented:
- The repository was updated to the current state of master.
- Added required patches:
  - 65e7721 - To work with factory layout Sercomm.
  - 3495cc5 - To work with factory U-boot Sercomm. 
- 4129951 - Changes:
  - OEM layouts
  - UART3 (for ZigBee)
  - nvmem-cells

MaxS0niX commented:
Ping @openwrt & @adschm, can you please help with the review? What needs to be fixed? What are the further actions?

FeelTheLemon commented:
Maybe that is because WIP label?
https://github.com/openwrt/openwrt/pull/4108#issuecomment-869202329

MaxS0niX commented:
> Maybe that is because WIP label?
> [#4108 (comment)](https://github.com/openwrt/openwrt/pull/4108#issuecomment-869202329)

Thank you, most likely this is the reason.

adschm commented on package/boot/uboot-envtools/files/ramips:
```diff
@@ -63,6 +63,9 @@ xiaomi,mi-router-ac2100|\
 xiaomi,redmi-router-ac2100)
 	ubootenv_add_uci_config "/dev/mtd1" "0x0" "0x1000" "0x20000"
 	;;
+beeline,smartbox-turbo-plus)
```
Alphabetic sorting ...

MaxS0niX replied:
Not sure if the sort is correct. Roughly corrected.

adschm commented on target/linux/ramips/dts/mt7621_beeline_smartbox-turbo-plus.dts:
```diff
@@ -0,0 +1,221 @@
+		* JTDI - GPIO#14 - Reset Button
+		* JTMS - GPIO#15 - RGB LED Green
+		* TCLK - GPIO#16 - RGB LED Red
+		*/
```
I don't think we need that comment here.

MaxS0niX replied:
I thought it would be better for clarification. Fixed.

adschm commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -226,6 +226,33 @@ define Device/asus_rt-n56u-b1
+  SERCOMM_KERNEL_OFFSET := 0x400100
+  SERCOMM_ROOTFS_OFFSET := 0x1000000
+  KERNEL := kernel-bin | append-dtb | lzma | loader-kernel | \
+    lzma | uImage lzma | sercomm-kernel
```
Hanging indent with one tab.

MaxS0niX replied:
Fixed.

MaxS0niX commented:
d1f2945 - Correction of the indicated errors.

MaxS0niX commented:
Is it possible to add a second sysupgrade image with a different partirions layout? Without adding a new device and partly another dts. I want to merge unused partitions through virtual memory (ubi-concat).

bobafetthotmail commented:
why a second sysupgrade image? just change the current dts.

Although having a working dual-partition setup with OpenWrt as well is imho better on this device, if users need to expand the storage they can just connect a USB drive and do an extroot

MaxS0niX commented:
> Although having a working dual-partition setup with OpenWrt as well is imho better on this device

Slot1 will not be able to boot, this requires that inside openwrt there are kernel & rootfs addresses for each slot. Is there a solution for this problem?

SopaDeMacaco-UmaDelicia commented:
@hauke, @mans0n, @981213 please show some love to this particular device and also #10118 #4770. Big work have been done to add support to these devices.

SopaDeMacaco-UmaDelicia commented:
A little bump and a question to the maintainers @hauke, @mans0n, @981213. Can we apply these [patches ](https://forum.openwrt.org/t/add-support-for-beeline-smartbox-turbo/99635/50) from @arinc9 to Turbo+ and #10118 pull requests before #10238 gets merged? The device maintainers fear that you will decline their PR since 2Gbps nat routing is kinda experimental and is not merged for the majority of devices.

csharper2005 commented:
> The device maintainers fear that you will decline their PR since 2Gbps nat routing is kinda experimental and is not merged for the majority of devices.

If I understand correctly the patch of mt7621.mk is required for the 5.15 kernel (current ramips testing kernel). Hence, we should change dtses after the https://github.com/openwrt/openwrt/pull/10238 is merged.

arinc9 commented:
Here's an idea. Apply my patches to this PR and #10118 and put a disclaimer requiring #10238 to be applied first. That'd add some incentive into having my PR and the device support PRs merged.

hauke commented on target/linux/ramips/dts/mt7621_beeline_smartbox-turbo-plus.dts:
```diff
@@ -0,0 +1,137 @@
+};
+
+&wlan_5g {
+	mediatek,mtd-eeprom = <&factory 0x8000>;
```
Are the MAC addresses set correctly for these interfaces?

hauke replied:
Please explain the MAC address usage in the commit description.

MaxS0niX replied:
> Are the MAC addresses set correctly for these interfaces?

https://forum.openwrt.org/t/add-support-for-beeline-smartbox-turbo/99635/21

MaxS0niX replied:
Added to commit.

hauke commented on target/linux/ramips/image/common-sercomm.mk:
```diff
@@ -85,3 +135,42 @@ define Device/sercomm_dxx
+define Device/uimage-lzma-loader-sercomm
+	LOADER_TYPE := bin
+	KERNEL := $(KERNEL_DTB) | loader-kernel | lzma -a0 | uImage lzma
+endef
```
Please add a new line

MaxS0niX replied:
Done.

hauke commented on target/linux/ramips/image/common-sercomm.mk:
```diff
@@ -64,6 +91,29 @@ define Build/sercomm-prepend-tagged-kernel
+	cat $@.hdrfactory $@.hdrkrn1 $@.hdrkrn2 $(IMAGE_KERNEL) $@ > $@.new
+	mv $@.new $@
+	rm -f $@.hdrfactory $@.hdrkrn1 $@.hdrkrn2
+endef
```
Please explain the image format in the commit description.

MaxS0niX replied:
Done.

hauke commented:
This pull request is still marked as draft is this intended?

csharper2005 commented on target/linux/ramips/dts/mt7621_beeline_smartbox-turbo.dts:
```diff
@@ -87,7 +37,7 @@
 		#size-cells = <1>;
 
 		partition@0 {
-			label = "u-boot";
```
I would like you not to change this generally accepted bootloader label.

MaxS0niX replied:
Specified a label as in the OEM firmware. Isn't that preferable?

csharper2005 replied:
There are two different aspects here:
1. In the commit "Add support for Device 1" you're trying to fix Device 2. I don't think this is a good idea. Would be better to fix another devices in the separate commits with description why do we need these fixes.
1. `u-boot` is a standard well known name for bootloader partition in OpenWrt. Like `rootfs`. So, I would prefer to keep this label.

csharper2005 commented on target/linux/ramips/dts/mt7621_sercomm_s3.dtsi:
```diff
@@ -0,0 +1,131 @@
+	};
+};
+
+&pcie1 { // MT7603EN
```
I would suggest use MAC address scheme (here and for 5g) from the newest firmware v.2.09 where 2g MAC is not equal to the LAN MAC.

MaxS0niX replied:
In the OEM firmware v.1.0.12, this was exactly the case. Does it make sense?

csharper2005 replied:
> Does it make sense?

Makes sense, since the vendor has fixed it in new firmware versions.
"Network nodes with multiple network interfaces, such as routers and multilayer switches, must have a unique MAC address for each NIC in the same network."

MaxS0niX replied:
But then it turns out that the manufacturer in vain indicated the MAC addresses of the wireless interfaces in their calibrations?

csharper2005 replied:
> But then it turns out that the manufacturer in vain indicated the MAC addresses of the wireless interfaces in their calibrations?

Manufacturer made a right decision not to use them in new fw releases because they aren't correct.

MaxS0niX replied:
Added to commit)

MaxS0niX commented:
> This pull request is still marked as draft is this intended?

Yes, in the draft because there was not quite the confidence in the correctness of the assembly recipes.

csharper2005 committed:
scripts: sercomm-kernel-header.py: improve compatibility

This improves compatibility with the elder stock firmwares of the
following devices, which have not yet been merged into OpenWrt:
 - Beeline SmartBox Pro
 - Beeline SmartBox Turbo+
 - WiFire S1500.NBN

Without this, OpenWrt factory image installation may fail.

Signed-off-by: Mikhail Zhilkin <csharper2005@gmail.com>
Signed-off-by: Maximilian Weinmann <x1@disroot.org>

csharper2005 reviewed (approved):
LGTM

MaxS0niX committed:
ramips: Add support for Beeline SmartBox TURBO+

This adds support for Beeline Smart Box TURBO+ (Serсomm S3 CQR) router.

Device specification
--------------------
SoC Type: MediaTek MT7621AT (880 MHz, 2 cores)
RAM (Nanya NT5CC64M16GP): 128 MiB
Flash (Macronix MX30LF1G18AC): 128 MiB
Wireless 2.4 GHz (MT7603EN): b/g/n, 2x2
Wireless 5 GHz (MT7615N): a/n/ac, 4x4
Ethernet: 5 ports - 5×GbE (WAN, LAN1-4)
USB ports: 1xUSB3.0
Buttons: 2 button (reset, wps)
LEDs: Red, Green, Blue
Zigbee (EFR32MG1B232GG): 3.0
Stock bootloader: U-Boot 1.1.3
Power: 12 VDC, 1.5 A

Installation (fw 2.0.9)
-----------------------
1.  Login to the web interface under SuperUser (root) credentials.
    Password: SDXXXXXXXXXX, where SDXXXXXXXXXX is serial number of the
    device written on the backplate stick.
2.  Navigate to Setting -> WAN. Add:
       Name - WAN1
       Connection Type - Static
       IP Address - 172.16.0.1
       Netmask - 255.255.255.0
    Save -> Apply. Set default: WAN1
3.  Enable SSH and HTTP on WAN. Setting -> Remote control. Add:
       Protocol - SSH
       Port - 22
       IP Address - 172.16.0.1
       Netmask - 255.255.255.0
       WAN Interface - WAN1
    Save ->Apply
    Add:
       Protocol - HTTP
       Port - 80
       IP Address - 172.16.0.1
       Netmask - 255.255.255.0
       WAN interface - WAN1
    Save -> Apply
4.  Set up your PC ethernet:
       Connection Type - Static
       IP Address - 172.16.0.2
       Netmask - 255.255.255.0
       Gateway - 172.16.0.1
5.  Connect PC using ethernet cable to the WAN port of the router
6.  Connect to the router using SSH shell under SuperUser account
7.  Make a mtd backup (optional, see related section)
8.  Change bootflag to Sercomm1 and reboot:
        printf 1 | dd bs=1 seek=7 count=1 of=/dev/mtdblock3
        reboot
9.  Login to the router web interface under admin account
10. Remove dots from the OpenWrt factory image filename
11. Update firmware via web using OpenWrt factory image

Revert to stock
---------------
Change bootflag to Sercomm1 in OpenWrt CLI and then reboot:
   printf 1 | dd bs=1 seek=7 count=1 of=/dev/mtdblock3

mtd backup
----------
1. Set up a tftp server (e.g. tftpd64 for windows)
2. Connect to a router using SSH shell and run the following commands:
      cd /tmp
      for i in 0 1 2 3 4 5 6 7 8 9 10; do nanddump -f mtd$i /dev/mtd$i; \
      tftp -l mtd$i -p 172.16.0.2; md5sum mtd$i >> mtd.md5; rm mtd$i; done
      tftp -l mtd.md5 -p 171.16.0.2

Recovery
--------
Use sercomm-recovery tool.
Link: https://github.com/danitool/sercomm-recovery

MAC Addresses (fw 2.0.9)
------------------------
+-----+------------+---------+
| use | address    | example |
+-----+------------+---------+
| LAN | label      | *:e8    |
| WAN | label + 1  | *:e9    |
| 2g  | label + 4  | *:ec    |
| 5g  | label + 5  | *:ed    |
+-----+------------+---------+
The label MAC address was found in Factory 0x21000

Factory image format
--------------------
+---+-------------------+-------------+--------------------+
| # | Offset            | Size        | Description        |
+---+-------------------+-------------+--------------------+
| 1 | 0x0               | 0x200       | Tag Header Factory |
| 2 | 0x200             | 0x100       | Tag Header Kernel1 |
| 3 | 0x300             | 0x100       | Tag Header Kernel2 |
| 4 | 0x400             | SIZE_KERNEL | Kernel             |
| 5 | 0x400+SIZE_KERNEL | SIZE_ROOTFS | RootFS(UBI)        |
+---+-------------------+-------------+--------------------+

Co-authored-by: Mikhail Zhilkin <csharper2005@gmail.com>
Signed-off-by: Maximilian Weinmann <x1@disroot.org>

SopaDeMacaco-UmaDelicia commented:
It's Alive!

hauke commented on target/linux/ramips/image/common-sercomm.mk:
```diff
@@ -16,6 +17,24 @@ define Build/sercomm-crypto
+		$@ $@.hdrkrnl)
+	cat $@.hdrfactory $@.hdrkrnl $(IMAGE_KERNEL).data $@ > $@.new
+	mv $@.new $@
+	rm -f $@.hdrfactory $@.hdrkrnl $(IMAGE_KERNEL).data
```
I do not like all these `call Build` commands. 

I think you should add the commands modifying the kernel image here:
````
  KERNEL := kernel-bin | append-dtb | lzma | loader-kernel | lzma -a0 | \
	uImage lzma | sercomm-kernel
````

Maybe line this:
````
  KERNEL := kernel-bin | append-dtb | lzma | loader-kernel | lzma -a0 | \
	uImage lzma | sercomm-kernel | sercomm-kernel-trim | sercomm-write-pid | sercomm-kernel-factory
````
(This is not directly working like this)

csharper2005 replied:
@hauke thanks for taking the time to look at these scripts. The essence of the problem that we need 2 different structures for the factory and sysupgrade images:
sysupgrade: `kernel header 1 | kernel `
factory:  `kernel header 1 | kernel header 2 | kernel `

Headers data depends on kernel data, ubi rootfs data and their offsets (which are different for the each Slot of the firmware)

This is correct for the sysupgrade (so, we can't touch this):
```
KERNEL := kernel-bin | append-dtb | lzma | loader-kernel | lzma -a0 | \
	uImage lzma | sercomm-kernel
```

We have to insert `kernel header 2` between `kernel header 1` and `kernel ` for the factory with correct checksums and so on. Also, factory image has a first header containing checksum of combined pure kernel + rootfs. That's why scripts are so complicated.

Perhaps I did not understood something. Could you explain your idea regarding kernel recipe in more detail, please?

csharper2005 replied:
Let's up this 2 yo PR. 🥇 

@hauke do we have any merge blockers except these `call Build` commands?

981213 replied:
Hi!

> This is correct for the sysupgrade (so, we can't touch this):
> 
> ```
> KERNEL := kernel-bin | append-dtb | lzma | loader-kernel | lzma -a0 | \
> 	uImage lzma | sercomm-kernel
> ```

You could drop the sercomm-kernel from KERNEL and apply it in factory.bin and sysupgrade.bin recipe separately. This way you can describe the recipe with steps, and 

> We have to insert `kernel header 2` between `kernel header 1` and `kernel ` for the factory with correct checksums and so on. Also, factory image has a first header containing checksum of combined pure kernel + rootfs. That's why scripts are so complicated.

This also eliminates the need to erase the existing `kernel header 1` before creating the factory image, as you are just adding the two header in factory.bin recipe.

csharper2005 replied:
Hi @981213! Thanks for your idea. What do you think about this?
```
diff --git a/target/linux/ramips/image/common-sercomm.mk b/target/linux/ramips/image/common-sercomm.mk
index 83bb2a742c..54f232fc6b 100644
--- a/target/linux/ramips/image/common-sercomm.mk
+++ b/target/linux/ramips/image/common-sercomm.mk
@@ -16,6 +16,23 @@ define Build/sercomm-crypto
 	rm -f $@.enc $@.key
 endef
 
+define Build/sercomm-factory-cqr
+	$(TOPDIR)/scripts/sercomm-pid.py \
+		--hw-version $(SERCOMM_HWVER) \
+		--hw-id $(SERCOMM_HWID) \
+		--sw-version $(SERCOMM_SWVER) \
+		--pid-file $@.fhdr
+	printf $$(stat -c%s $(IMAGE_KERNEL)) | \
+		dd seek=$$((0x70)) of=$@.fhdr bs=1 conv=notrunc 2>/dev/null
+	printf $$(($$(stat -c%s $@)-$$(stat -c%s $(IMAGE_KERNEL))-$$((0x200)))) | \
+		dd seek=$$((0x80)) of=$@.fhdr bs=1 conv=notrunc 2>/dev/null
+	dd bs=$$((0x200)) skip=1 if=$@ conv=notrunc 2>/dev/null | \
+		$(MKHASH) md5 | awk '{print $$1}' | tr -d '\n' | \
+		dd seek=$$((0x1e0)) of=$@.fhdr bs=1 conv=notrunc 2>/dev/null
+	dd if=$@ >> $@.fhdr 2>/dev/null
+	mv $@.fhdr $@
+endef
+
 define Build/sercomm-kernel
 	$(TOPDIR)/scripts/sercomm-kernel-header.py \
 		--kernel-image $@ \
@@ -26,6 +43,22 @@ define Build/sercomm-kernel
 	mv $@.hdr $@
 endef
 
+define Build/sercomm-kernel-factory
+	$(TOPDIR)/scripts/sercomm-kernel-header.py \
+		--kernel-image $@ \
+		--kernel-offset $(SERCOMM_KERNEL_OFFSET) \
+		--rootfs-offset $(SERCOMM_ROOTFS_OFFSET) \
+		--output-header $@.khdr1
+	$(TOPDIR)/scripts/sercomm-kernel-header.py \
+		--kernel-image $@ \
+		--kernel-offset $(SERCOMM_KERNEL2_OFFSET) \
+		--rootfs-offset $(SERCOMM_ROOTFS2_OFFSET) \
+		--output-header $@.khdr2
+	cat $@.khdr1 $@.khdr2 > $@.khdr
+	dd if=$@ >> $@.khdr 2>/dev/null
+	mv $@.khdr $@
+endef
+
 define Build/sercomm-part-tag
 	$(call Build/sercomm-part-tag-common,$(word 1,$(1)) $@)
 endef
@@ -64,6 +97,28 @@ define Build/sercomm-prepend-tagged-kernel
 	mv $(IMAGE_KERNEL).tagged $@
 endef
 
+define Device/sercomm_cxx
+  $(Device/dsa-migration)
+  BLOCKSIZE := 128k
+  PAGESIZE := 2048
+  UBINIZE_OPTS := -E 5
+  LOADER_TYPE := bin
+  KERNEL := kernel-bin | append-dtb | lzma | loader-kernel | lzma -a0 | \
+	uImage lzma
+  IMAGES += factory.img
+  IMAGE/sysupgrade.bin := append-kernel | sercomm-kernel | \
+	sysupgrade-tar kernel=$$$$@ | append-metadata
+  KERNEL_SIZE := 6144k
+  KERNEL_LOADADDR := 0x81001000
+  LZMA_TEXT_START := 0x82800000
+  SERCOMM_KERNEL_OFFSET := 0x400100
+  SERCOMM_ROOTFS_OFFSET := 0x1000000
+  SERCOMM_KERNEL2_OFFSET := 0xa00100
+  SERCOMM_ROOTFS2_OFFSET := 0x3000000
+  IMAGE/factory.img := append-kernel | sercomm-kernel-factory | \
+	append-ubi | sercomm-factory-cqr
+endef
+
 define Device/sercomm_dxx
   $(Device/dsa-migration)
   BLOCKSIZE := 128k

```

981213 replied:
> What do you think about this?

Yeah, this one is better :D

DragonBluep commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -378,6 +378,19 @@ define Device/beeline_smartbox-turbo
+  SERCOMM_SWVER := 2010
+  DEVICE_VENDOR := Beeline
+  DEVICE_MODEL := SmartBox TURBO+
+  DEVICE_PACKAGES := kmod-mt7603 kmod-mt7615e kmod-mt7615-firmware \
```
`kmod-mt7615e` can be removed now. Reference: https://github.com/openwrt/openwrt/commit/9a07895729a9701a673951a31ceb71f16f0a6fad

MaxS0niX replied:
Fix

SopaDeMacaco-UmaDelicia commented:
It's not Alive!

csharper2005 commented:
> It's not Alive!

This router seems is being under a spell. :)

SopaDeMacaco-UmaDelicia commented:
@hauke can you check this PR? Talk to us, this PR is long overdue, we don't know what openwrt maintainers want from us.

nabOl5 commented:
Oh, at last it is moving! Thank a lot!

DragonBluep commented on target/linux/ramips/image/common-sercomm.mk:
```diff
@@ -64,24 +113,45 @@ define Build/sercomm-prepend-tagged-kernel
+
+define Device/sercomm_cxx
+  $(Device/sercomm)
+  KERNEL_SIZE := 6144k
```
I'd like to apply these changes:
```
--- a/target/linux/ramips/image/common-sercomm.mk
+++ b/target/linux/ramips/image/common-sercomm.mk
@@ -119,26 +119,22 @@ define Build/sercomm-write-pid
 endef
 
 define Device/sercomm
-  $(Device/dsa-migration)
-  BLOCKSIZE := 128k
-  PAGESIZE := 2048
-  UBINIZE_OPTS := -E 5
+  $(Device/nand)
+  KERNEL_SIZE := 6144k
   LOADER_TYPE := bin
+  KERNEL_LOADADDR := 0x81001000
+  LZMA_TEXT_START := 0x82800000
+  SERCOMM_KERNEL_OFFSET := 0x400100
+  SERCOMM_ROOTFS_OFFSET := 0x1000000
   KERNEL := kernel-bin | append-dtb | lzma | loader-kernel | lzma -a0 | \
 	uImage lzma | sercomm-kernel
   KERNEL_INITRAMFS := kernel-bin | append-dtb | lzma | loader-kernel | \
 	lzma -a0 | uImage lzma
   IMAGES += factory.img
-  IMAGE/sysupgrade.bin := sysupgrade-tar | append-metadata
 endef
 
 define Device/sercomm_cxx
   $(Device/sercomm)
-  KERNEL_SIZE := 6144k
-  KERNEL_LOADADDR := 0x81001000
-  LZMA_TEXT_START := 0x82800000
-  SERCOMM_KERNEL_OFFSET := 0x400100
-  SERCOMM_ROOTFS_OFFSET := 0x1000000
   SERCOMM_KERNEL2_OFFSET := 0xa00100
   SERCOMM_ROOTFS2_OFFSET := 0x3000000
   IMAGE/factory.img := append-ubi | check-size | sercomm-factory-cqr
@@ -146,11 +142,6 @@ endef
 
 define Device/sercomm_dxx
   $(Device/sercomm)
-  KERNEL_SIZE := 6144k
-  KERNEL_LOADADDR := 0x81001000
-  LZMA_TEXT_START := 0x82800000
-  SERCOMM_KERNEL_OFFSET := 0x400100
-  SERCOMM_ROOTFS_OFFSET := 0x1000000
   IMAGE/factory.img := append-ubi | check-size | \
 	sercomm-part-tag rootfs | sercomm-prepend-tagged-kernel kernel | \
 	gzip | sercomm-payload | sercomm-crypto

```

MaxS0niX replied:
New changes take into account the wish)

DragonBluep commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -408,6 +408,19 @@ define Device/beeline_smartbox-turbo
+  DEVICE_VENDOR := Beeline
+  DEVICE_MODEL := SmartBox TURBO+
+  DEVICE_PACKAGES := kmod-mt7603 kmod-mt7615-firmware kmod-usb3 \
+    uboot-envtools
```
`uboot-envtools` can be removed as it's a default package now.

MaxS0niX replied:
Done.

nabOl5 replied:
`Tested-by: Oleg Nabatchikov <2scorpio@inbox.ru>`

SopaDeMacaco-UmaDelicia commented:
gentle bump @981213 @hauke

nabOl5 commented:
Up for review. :)

SopaDeMacaco-UmaDelicia commented:
And here we go again, reviewers will be back in half a year, as always.

csharper2005 commented:
> And here we go again, reviewers will be back in half a year, as always.

I understand your impatience due to the long review. I also really want this router to get official OpenWrt support. 

At the same time, it is Open Source. There are no commitments or SLAs here. IOW, no one owes nothing to nobody. That's OK.

Nevertheless, I hope @adschm, @hauke, @981213 or other committer find time soon to check / merge this PR.

981213 commented:
Merged into my staging tree at: https://git.openwrt.org/?p=openwrt/staging/981213.git;a=shortlog;h=refs/heads/main
Thanks!

csharper2005 commented:
> Merged into my staging tree at: https://git.openwrt.org/?p=openwrt/staging/981213.git;a=shortlog;h=refs/heads/main Thanks!

Thanks a lot for your help! How often your staging tree syncs with the `main` one?
